singularity-eos/eos/modifiers/zsplit_eos.hpp

```diff
@@ -0,0 +1,274 @@
+// reserved by Triad National Security, LLC, and the U.S. Department of
+// Energy/National Nuclear Security Administration. The Government is
+// granted for itself and others acting on its behalf a nonexclusive,
+// paid-up, irrevocable worldwide license in this material to reproduce,
+// prepare derivative works, distribute copies to the public, perform
+// publicly and display publicly, and to permit others to do so.
+//------------------------------------------------------------------------------
+
+#ifndef _SINGULARITY_EOS_EOS_ZSPLIT_EOS_
+#define _SINGULARITY_EOS_EOS_ZSPLIT_EOS_
+
+#include <cmath>
+#include <utility>
+
+#include <ports-of-call/portability.hpp>
+#include <ports-of-call/portable_errors.hpp>
+#include <singularity-eos/base/constants.hpp>
+#include <singularity-eos/base/eos_error.hpp>
+#include <singularity-eos/base/robust_utils.hpp>
+#include <singularity-eos/eos/eos_base.hpp>
+
+namespace singularity {
+
+using namespace eos_base;
+
+enum class ZSplitComponent { Electrons, Ions };
+
+// TODO(JMM): Separate type? Or runtime?
+template <ZSplitComponent ztype, typename T>
+class ZSplit : public EosBase<ZSplit<ztype, T>> {
+ public:
+  using BaseType = T;
+  static constexpr const ZSplitComponent zsplit_type = ztype;
+  SG_ADD_BASE_CLASS_USINGS(ZSplit<ztype, T>);
+
+  static std::string EosType() {
+    std::string ts = (ztype == ZSplitComponent::Electrons) ? "Electrons, " : "Ions, ";
+    return std::string("ZSplit<") + ts + T::EosType() + std::string(">");
+  }
+  static std::string EosPyType() {
+    std::string ts = (ztype == ZSplitComponent::Electrons) ? "Electrons" : "Ions";
+    return std::string("ZSplit") + ts + T::EosPyType();
+  }
+
+  ZSplit() = default;
+  PORTABLE_INLINE_FUNCTION
+  ZSplit(T &&t) : t_(std::forward<T>(t)) {}
+
+  PORTABLE_INLINE_FUNCTION void CheckParams() const { t_.CheckParams(); }
+  auto GetOnDevice() { return ZSplit<ztype, T>(t_.GetOnDevice()); }
+  inline void Finalize() { t_.Finalize(); }
+
+  template <typename Indexer_t = Real *>
+  PORTABLE_FUNCTION Real TemperatureFromDensityInternalEnergy(
+      const Real rho, const Real sie, Indexer_t &&lambda = nullptr) const {
+    const Real iscale = GetInvScale_(lambda);
+    return t_.TemperatureFromDensityInternalEnergy(rho, sie * iscale, lambda);
+  }
+  template <typename Indexer_t = Real *>
+  PORTABLE_FUNCTION Real InternalEnergyFromDensityTemperature(
+      const Real rho, const Real temperature, Indexer_t &&lambda = nullptr) const {
+    const Real scale = GetScale_(lambda);
+    const Real et = t_.InternalEnergyFromDensityTemperature(rho, temperature, lambda);
+    return scale * et;
+  }
+
+  template <typename Indexer_t = Real *>
+  PORTABLE_FUNCTION Real PressureFromDensityTemperature(
+      const Real rho, const Real temperature, Indexer_t &&lambda = nullptr) const {
+    const Real scale = GetScale_(lambda);
+    const Real Pt = t_.PressureFromDensityTemperature(rho, temperature, lambda);
```

Comment:
This model really isn't compatible with tension---it's derived from an ideal gas, after all.

I think we have to hope that in tension, the ionization state is approximately zero, so that ion and total pressure are both negative. If you have tension the zsplit will make the electron pressure negative too, but that shouldn't matter if we're not ionized.

Physically, that seems like a reasonable assumption... after all, if you're in tension, you're presumably cold enough that you're not ionized. But perhaps this means we should floor the electron pressure.

Reply:
I've seen wonky things with ionization models (usually for gasses) where they claim to be ionized at low temperatures. A floor wouldn't be unreasonable

Reply:
Gross.